### PR TITLE
Replacing TS source mapper

### PR DESCRIPTION
**Pull request by ahmad-farid**
- Replacing the current source mapper of NTVS to use the TS Source Map Reader used in VS.
- Since this is still an experimental branch, keeping the original source mapper files for reference.
- The newly added files are part of the TS Source Map Reader or its requirements. Those were not altered and copied as is.

### CODE REVIEW

File: Nodejs/Product/Nodejs/Debugger/BinaryDeserializer.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.IO;
+using TypeScriptSourceMapReader;
+
+namespace Microsoft.NodejsTools.Debugger {
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/BinaryDeserializer.cs]: Make IDisposable to clean up memorystream and binaryreader

File: Nodejs/Product/Nodejs/Debugger/BinarySerializer.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+using System.IO;
+using System.Linq;
+using TypeScriptSourceMapReader;
+
+namespace Microsoft.NodejsTools.Debugger {
+    /// <summary>
+    /// Custom serializer to convert the different types to byte[]
+    /// The serialization and deserialization needs to be always in sync
+    /// </summary>
+    internal class BinarySerializer
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/BinarySerializer.cs]: Make IDisposable to clean up memorystream and binarywriter

File: Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs [hunk 3]
```diff
@@ -55,18 +59,53 @@ sealed class NodeBreakpoint {
...
+             return Target;
+        }
 
-            if (mapper != null &&
-                mapper.MapToJavaScript(Target.FileName, Target.Line, Target.Column, out javaScriptFileName, out javaScriptLine, out javaScriptColumn)) {
-                return new FilePosition(javaScriptFileName, javaScriptLine, javaScriptColumn);
-            }
-
-            return Target;
+        string FindSourceMapFile(string jsFileName) {
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs]: mark private/internal

File: Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs [hunk 4]
```diff
@@ -55,18 +59,53 @@ sealed class NodeBreakpoint {
...
+            string sourceMapFilename = null;
+
+            if (File.Exists(jsFileName)) {
+                string[] contents = File.ReadAllLines(jsFileName);
+                const string marker = "# sourceMappingURL=";
+                int markerStart;
+                string markerLine = contents.Reverse().FirstOrDefault(x => x.IndexOf(marker, StringComparison.Ordinal) != -1);
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs]: isn't a sourceMappingUrl by definition the last line?

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 5]
```diff
@@ -44,9 +46,8 @@ sealed class NodeDebugger : IDisposable {
         private readonly ExceptionHandler _exceptionHandler;
         private readonly Dictionary<string, NodeModule> _modules = new Dictionary<string, NodeModule>(StringComparer.OrdinalIgnoreCase);
         private readonly EvaluationResultFactory _resultFactory;
-        private readonly SourceMapper _sourceMapper;
         private readonly Dictionary<int, NodeThread> _threads = new Dictionary<int, NodeThread>();
-        private readonly TimeSpan _timeout = TimeSpan.FromSeconds(5);
+        private readonly TimeSpan _timeout = TimeSpan.FromSeconds(15);
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: why did you change this?

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 6]
```diff
@@ -810,6 +861,10 @@ public NodeDebugger(Uri debuggerEndpointUri, int id)
             }
         }
 
+        private void LoadDecodedSourceMap() {
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: Remove or implement

File: Nodejs/Product/Nodejs/Debugger/TypeScriptRuntime.cs [hunk 7]
```diff
@@ -0,0 +1,206 @@
+﻿using Microsoft.VisualStudio.Debugger;
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/TypeScriptRuntime.cs]: System comes first

File: Nodejs/Product/Nodejs/Debugger/BinaryDeserializer.cs (hunk 1 above)
Comment (ahmad-farid) [inline Nodejs/Product/Nodejs/Debugger/BinaryDeserializer.cs]: As we spoke today, this, as well as the other newly added files, were imported as is. It's a valid point, but we should change them in the original source location where they'd be shared from.

File: Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs (hunk 4 above)
Comment (ahmad-farid) [inline Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs]: This the the same logic used by the main SourceMap Reader. @sheetalkamat could shed more light on this.

File: Nodejs/Product/Nodejs/Debugger/BinarySerializer.cs (hunk 2 above)
Comment (ahmad-farid) [inline Nodejs/Product/Nodejs/Debugger/BinarySerializer.cs]: Same as above. As we discussed, copied as is.

File: Nodejs/Product/Nodejs/Debugger/TypeScriptRuntime.cs (hunk 7 above)
Comment (ahmad-farid) [inline Nodejs/Product/Nodejs/Debugger/TypeScriptRuntime.cs]: Same as above. As we discussed, copied as is.

File: Nodejs/NodejsTools.sln [hunk 8]
```diff
@@ -71,120 +71,196 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "AnalysisDriver", "Tests\Ana
...
+EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
 		Debug|x86 = Debug|x86
+		Release|Any CPU = Release|Any CPU
 		Release|x86 = Release|x86
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{32EC5259-98DA-40CA-9E2D-1B1B2E966F88}.Debug|Any CPU.ActiveCfg = Debug|x86
```
Comment (mousetraps) [inline Nodejs/NodejsTools.sln]: remove the new extraneous configuration

File: Nodejs/Product/Nodejs/Debugger/BinaryDeserializer.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
+﻿using System;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/BinaryDeserializer.cs]: Add copyright headers to all new files

File: Nodejs/Product/Nodejs/Debugger/BinarySerializer.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
+﻿using System;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/BinarySerializer.cs]: Can you add the set of files that's copied from elsewhere to a separate folder and add comments to make it clearer that they're shared components?

File: Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs [hunk 11]
```diff
@@ -55,18 +59,53 @@ sealed class NodeBreakpoint {
...
-        public FilePosition GetPosition(SourceMapper mapper) {
+        public FilePosition GetPosition(string javaScriptFileName) {
             // Checks whether source map is available
-            string javaScriptFileName;
+
             int javaScriptLine;
             int javaScriptColumn;
+            DecodedSourceMap decodedSourceMap = null;
+            DkmTextSpan? jsSpan = null;
+            string sourceMapFilename;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs]: define these as close to their instantiations as possible

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 12]
```diff
@@ -777,6 +784,29 @@ public NodeDebugger(Uri debuggerEndpointUri, int id)
...
+                        if (!File.Exists(sourceMapFilename)) {
+                            sourceMapFilename = Path.Combine(Path.GetDirectoryName(jsFileName) ?? string.Empty, Path.GetFileName(sourceMapFilename));
+                        }
+                    } catch (ArgumentException) {
+                    } catch (PathTooLongException) {
+                    }
+                }
+            }
+            return sourceMapFilename;
         }
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: duplicate method

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 13]
```diff
@@ -1218,11 +1272,35 @@ public NodeDebugger(Uri debuggerEndpointUri, int id)
...
+
+                if (DecodedSourceMaps.ContainsKey(sourceMapFilename)) {
+                    decodedSourceMap = DecodedSourceMaps[sourceMapFilename];
+                } else {
+                    decodedSourceMap = SourceMapReader.LoadSourceMap(javaScriptFileName, sourceMapFilename);
+                    DecodedSourceMaps.Add(sourceMapFilename, decodedSourceMap);
+                    _sourceMapFiles.Add(javaScriptFileName, sourceMapFilename);
+                }
+
+                decodedSourceMap.MapJsSourcePosition(new DkmTextSpan((int)line + 1, (int)line + 1, (int)column + 1, (int)column + 1), out sourceInfo, out name);
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: define name above this file since you don't actually use it for anything other than the output arg

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 14]
```diff
@@ -1218,11 +1272,35 @@ public NodeDebugger(Uri debuggerEndpointUri, int id)
...
                 column = stackFrame.Column;
             }
-            string originalFileName = SourceMapper.GetOriginalFileName(javaScriptFileName, line, column);
+
+            SourceMapSourceInfo sourceInfo = null;
+            string name;
+
+            string sourceMapFilename = FindSourceMapFile(javaScriptFileName);
+
+            DecodedSourceMap decodedSourceMap;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: move this into the if

File: Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    <GenerateAssemblyAttribute>false</GenerateAssemblyAttribute>
+    <ConvertLocalizableResourcesToDefaultCulture>True</ConvertLocalizableResourcesToDefaultCulture>
+    <UICulture>en</UICulture>
+    <GenerateLCE>true</GenerateLCE>
+    <SccProjectName>SAK</SccProjectName>
+    <SccLocalPath>SAK</SccLocalPath>
+    <SccAuxPath>SAK</SccAuxPath>
+    <SccProvider>SAK</SccProvider>
+    <OutputPath>$(BinariesDirectory)\bin\$(BuildArchitecture)\typescript\TypeScriptSourceMapReader</OutputPath>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
```
Comment (mousetraps) [inline Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj]: This won't work on Dev11

Comment (mousetraps) [inline Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj]: Just filed an issue here to address this later. https://github.com/Microsoft/nodejstools/issues/444

File: Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs [hunk 16]
```diff
@@ -55,18 +59,53 @@ sealed class NodeBreakpoint {
...
+                if (markerLine != null && (markerStart = markerLine.IndexOf(marker, StringComparison.Ordinal)) != -1) {
+                    sourceMapFilename = markerLine.Substring(markerStart + marker.Length).Trim();
+
+                    try {
+                        if (!File.Exists(sourceMapFilename)) {
+                            sourceMapFilename = Path.Combine(Path.GetDirectoryName(jsFileName) ?? string.Empty, Path.GetFileName(sourceMapFilename));
+                        }
+                    } catch (ArgumentException) {
+                    } catch (PathTooLongException) {
+                    }
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs]: when would an ArgumentException occur? can you add a comment?

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 17]
```diff
@@ -29,7 +29,9 @@
 using Microsoft.NodejsTools.Debugger.Serialization;
 using Microsoft.NodejsTools.SourceMapping;
 using Microsoft.VisualStudioTools;
-
+using Microsoft.VisualStudio.Debugger.Symbols;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: this should be sorted before VisualStudioTools

File: Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs [hunk 18]
```diff
@@ -55,18 +59,53 @@ sealed class NodeBreakpoint {
...
-            return Target;
+        private string FindSourceMapFile(string jsFileName) {
+            string sourceMapFilename = null;
+
+            if (File.Exists(jsFileName)) {
+                string[] contents = File.ReadAllLines(jsFileName);
+                const string marker = "# sourceMappingURL=";
+                int markerStart;
+                string markerLine = contents.Reverse().FirstOrDefault(x => x.IndexOf(marker, StringComparison.Ordinal) != -1);
+                if (markerLine != null && (markerStart = markerLine.IndexOf(marker, StringComparison.Ordinal)) != -1) {
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs]: can you define markerStart above rather than inline?

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 19]
```diff
@@ -789,14 +819,35 @@ public NodeDebugger(Uri debuggerEndpointUri, int id)
                 int line = stackFrame.Line;
                 int column = stackFrame.Column;
                 string functionName = stackFrame.FunctionName;
+                int lineEnd = stackFrame.EndLine;
+
+                string name = "";
+                //int tsLine = -1, tsColumn = -1;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: extraneous comment

File: Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs [hunk 20]
```diff
@@ -789,14 +819,35 @@ public NodeDebugger(Uri debuggerEndpointUri, int id)
...
+                    string jsFileName = module.JavaScriptFileName;
+                    string sourceMapFilename = FindSourceMapFile(jsFileName);
+                    SourceMapSourceInfo sourceInfo = null;
+                    if (!string.IsNullOrEmpty(sourceMapFilename)) {
+
+                        if (DecodedSourceMaps.ContainsKey(sourceMapFilename)) {
+                            decodedSourceMap = DecodedSourceMaps[sourceMapFilename];
+                        } else {
+                            decodedSourceMap = SourceMapReader.LoadSourceMap(jsFileName, sourceMapFilename);
+                            DecodedSourceMaps.Add(sourceMapFilename, decodedSourceMap);
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugger.cs]: is there ever a case where decodedSourceMap is null and shouldn't be added?

File: Nodejs/Product/Nodejs/Nodejs.csproj [hunk 21]
```diff
@@ -1578,6 +1587,9 @@
     <RegisterOutputPackage>true</RegisterOutputPackage>
     <RegisterWithCodebase>true</RegisterWithCodebase>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
+    <DefineConstants>DEBUG;TRACE;;FEATURE_AZURE_REMOTE_DEBUG;DEV14;DEV14_OR_LATER;DEV12_OR_LATER;DEV11_OR_LATER;NODEJS_TOOLS;NTVS_FEATURE_INTERACTIVEWINDOW</DefineConstants>
+  </PropertyGroup>
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: unnecessary

File: Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj [hunk 22]
```diff
@@ -0,0 +1,106 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="dogfood" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <ProductVersion>9.0.30729</ProductVersion>
+    <SchemaVersion>2.0</SchemaVersion>
+    <ProjectGuid>{61A8D7EF-B83E-4C21-9479-BA832357E42C}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>TypeScriptSourceMapReader</RootNamespace>
+    <AssemblyName>TypeScriptSourceMapReader</AssemblyName>
```
Comment (mousetraps) [inline Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj]: Also add this assembly to the installer and to our BuildRelease script for signing

File: Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+    <Compile Include="SourceMapReader.cs" />
+    <Compile Include="SourceMapReaderException.cs" />
+    <Compile Include="SourceMapSourceInfo.cs" />
+    <Compile Include="SourceMapSpan.cs" />
+    <Compile Include="SourceMapSpanMapping.cs" />
+    <Compile Include="SourceMapTextReader.cs" />
+    <Compile Include="SourceMapUrlHelper.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="FinalPublicKey.snk" />
```
Comment (mousetraps) [inline Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj]: is this necessary?

File: Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+    <ErrorReport>prompt</ErrorReport>
+    <DebugType>full</DebugType>
+  </PropertyGroup>
+  <Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />
+  <ItemGroup>
+    <ManagedPreprocessFile Include="$(AssemblyAttributesFileName)">
+      <OutputFileExtension>.cs</OutputFileExtension>
+      <SuppressLineNumbers>true</SuppressLineNumbers>
+      <Defines>CSC_INVOKED=1;PPTARGET_CS=1</Defines>
+    </ManagedPreprocessFile>
```
Comment (mousetraps) [inline Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj]: what's this for?

File: Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    <UICulture>en</UICulture>
+    <GenerateLCE>true</GenerateLCE>
+    <SccProjectName>SAK</SccProjectName>
+    <SccLocalPath>SAK</SccLocalPath>
+    <SccAuxPath>SAK</SccAuxPath>
+    <SccProvider>SAK</SccProvider>
+    <OutputPath>$(BinariesDirectory)\bin\$(BuildArchitecture)\typescript\TypeScriptSourceMapReader</OutputPath>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <DebuggerAPIRefPath>$(InternalAPIsPath)\debugger\ref\v4.5\$(InterApiType)</DebuggerAPIRefPath>
+    <NewtonsoftJsonPath>$(EnlistmentRootPath)\externalapis\AspWebStack\Newtonsoft.Json.6.0.4\lib\net45</NewtonsoftJsonPath>
```
Comment (mousetraps) [inline Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj]: these look like internal paths

File: Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+    <Compile Include="SourceMapSourceInfo.cs" />
+    <Compile Include="SourceMapSpan.cs" />
+    <Compile Include="SourceMapSpanMapping.cs" />
+    <Compile Include="SourceMapTextReader.cs" />
+    <Compile Include="SourceMapUrlHelper.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="FinalPublicKey.snk" />
+  </ItemGroup>
+  <Import Project="..\ProjectAfter.settings" />
```
Comment (mousetraps) [inline Nodejs/Product/TypeScriptSourceMapReader/TypeScriptSourceMapReader.csproj]: There are a bunch of things we can simplify here - go ahead and take a look at Npm.csproj to for an example of what's necessary and what's not: https://github.com/Microsoft/nodejstools/blob/master/Nodejs/Product/Npm/Npm.csproj

File: Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs (hunk 16 above)
Comment (ahmad-farid) [inline Nodejs/Product/Nodejs/Debugger/NodeBreakpoint.cs]: @sheetalkamat - do you know? I copied this section from the main TS code. Generally, this exception might be thrown when illegal character are present in the path.

**Comment by mousetraps** [comment]
Merging into source-mapping branch, will address any remaining comments there and make a PR against master.
